ros/src/computing/perception/detection/lidar_detector/packages/lidar_naive_l_shape_detect/nodes/lidar_naive_l_shape_detect/lidar_naive_l_shape_detect.cpp

```diff
@@ -0,0 +1,305 @@
+    int offset_init_x = roi_m_ * pic_scale_ / 2 - init_pic_x;
+    int offset_init_y = roi_m_ * pic_scale_ / 2 - init_pic_y;
+
+    int num_points = cloud.size();
+    std::vector<cv::Point> point_vec(num_points);
+    std::vector<cv::Point2f> pc_points(4);
+
+    // init variables
+    float min_mx = 0;
+    float min_my = 0;
+    float max_mx = 0;
+    float max_my = 0;
+    float min_m = std::numeric_limits<float>::max();
+    float max_m = std::numeric_limits<float>::min();
+    float max_z = std::numeric_limits<float>::min();
+
+    for (int i_point = 0; i_point < num_points; i_point++)
+    {
+      float p_x = cloud[i_point].x;
+      float p_y = cloud[i_point].y;
+      float p_z = cloud[i_point].z;
+      // cast (roi_m_/2 < x,y < roi_m_/2) into (0 < x,y < roi_m_)
+      float roi_x = p_x + roi_m_ / 2;
+      float roi_y = p_y + roi_m_ / 2;
+      // cast (roi_m_)m*(roi_m_)m into  pic_scale_
+      int x = floor(roi_x * pic_scale_);
+      int y = floor(roi_y * pic_scale_);
+      // cast into image coordinate
+      int pic_x = x;
+      int pic_y = pic_scale_ * roi_m_ - y;
+      // offset so that the object would be locate at the center
+      int offset_x = pic_x + offset_init_x;
+      int offset_y = pic_y + offset_init_y;
+
+      // Make sure points are inside the image size
+      if (offset_x > (pic_scale_ * roi_m_) || offset_x < 0 || offset_y < 0 || offset_y > (pic_scale_ * roi_m_))
+      {
+        continue;
+      }
+      // cast the pointcloud into cv::mat
+      m.at<uchar>(offset_y, offset_x) = 255;
+      point_vec[i_point] = cv::Point(offset_x, offset_y);
+      // calculate min and max slope for x1, x3(edge points)
+      float delta_m = p_y / p_x;
+      if (delta_m < min_m)
+      {
+        min_m = delta_m;
+        min_mx = p_x;
+        min_my = p_y;
+      }
+
+      if (delta_m > max_m)
+      {
+        max_m = delta_m;
+        max_mx = p_x;
+        max_my = p_y;
+      }
+
+      // get maxZ
+      if (p_z > max_z)
+      {
+        max_z = p_z;
+      }
+    }
+
+    if (max_m == std::numeric_limits<float>::min() || min_m == std::numeric_limits<float>::max() ||
+        max_z == std::numeric_limits<float>::min())
+    {
+      continue;
+    }
+    // L shape fitting parameters
+    float x_dist = max_mx - min_mx;
+    float y_dist = max_my - min_my;
+    float slope_dist = sqrt(x_dist * x_dist + y_dist * y_dist);
+    float slope = (max_my - min_my) / (max_mx - min_mx);
+
+    // random variable
+    std::mt19937_64 mt;
+    mt.seed(in_object_array.header.stamp.toSec());
+    // mt.seed(0);
+    std::uniform_int_distribution<> rand_points(0, num_points - 1);
+
+    // start l shape fitting for car like object
+    if (slope_dist > slope_dist_thres_ && num_points > num_points_thres_)
+    {
+      float max_dist = 0;
+      float max_dx = 0;
+      float max_dy = 0;
+
+      // num_random_points, get max distance
+      for (int i = 0; i < ram_points_; i++)
```

Comment:
Is the index _i_ used in this loop? It seems that there is no need to be a loop.